optuna/cli.py

```diff
@@ -886,48 +896,184 @@ def take_action(self, parsed_args: Namespace) -> int:
         return 0
 
 
-class _OptunaApp(App):
-    def __init__(self) -> None:
-
-        super().__init__(
-            description="",
-            version=optuna.__version__,
-            command_manager=CommandManager("optuna.command"),
-            deferred_help=True,
+_COMMANDS: Dict[str, Type[_BaseCommand]] = {
+    "create-study": _CreateStudy,
+    "delete-study": _DeleteStudy,
+    "study set-user-attr": _StudySetUserAttribute,
+    "studies": _Studies,
+    "trials": _Trials,
+    "best-trial": _BestTrial,
+    "best-trials": _BestTrials,
+    "study optimize": _StudyOptimize,
+    "storage upgrade": _StorageUpgrade,
+    "ask": _Ask,
+    "tell": _Tell,
+}
+
+
+def _add_common_arguments(parser: ArgumentParser) -> ArgumentParser:
+    parser.add_argument("--storage", default=None, help="DB URL. (e.g. sqlite:///example.db)")
+    verbose_group = parser.add_mutually_exclusive_group()
+    verbose_group.add_argument(
+        "-v",
+        "--verbose",
+        action="count",
+        dest="verbose_level",
+        default=1,
+        help="Increase verbosity of output. Can be repeated.",
+    )
+    verbose_group.add_argument(
+        "-q",
+        "--quiet",
+        action="store_const",
+        dest="verbose_level",
+        const=0,
+        help="Suppress output except warnings and errors.",
+    )
+    parser.add_argument(
+        "--log-file",
+        action="store",
+        default=None,
+        help="Specify a file to log output. Disabled by default.",
+    )
+    parser.add_argument(
+        "--debug",
+        default=False,
+        action="store_true",
+        help="Show tracebacks on errors.",
+    )
+    return parser
+
+
+def _add_commands(
+    main_parser: ArgumentParser, parent_parser: ArgumentParser
+) -> Tuple[ArgumentParser, Dict[str, ArgumentParser]]:
+    subparsers = main_parser.add_subparsers()
+    command_name_to_subparser = {}
+
+    for (command_name, command_type) in _COMMANDS.items():
+        command = command_type()
+        subparser = subparsers.add_parser(
+            command_name, parents=[parent_parser], help=inspect.getdoc(command_type)
         )
+        subparser = command.add_arguments(subparser)
+        subparser.set_defaults(handler=command.take_action)
+        command_name_to_subparser[command_name] = subparser
+
+    def _print_help(args: Namespace) -> None:
+        main_parser.print_help()
+
+    subparsers.add_parser("help", help="Show help message and exit.").set_defaults(
+        handler=_print_help
+    )
+    return main_parser, command_name_to_subparser
```

Comment:
I wonder if we don't need to return `main_parser` because the argument is also modified. Is there any reason to do this?

Reply:
I confirmed its redundant code. Thanking for pointing that out. Fixed.